### PR TITLE
Close LGVISIUM-52: Remove grouping by page

### PR DESCRIPTION
**Pull request by dcleres**
Remove the page based logic in the code. The `prediction.json`  has now the following shape: 

```json
{
    "685256002-bp.pdf": {
        "language": "de",
        "metadata": {
            "coordinates": null
        },
        "layers": [
            {
                "material_description": {
                    "text": "grauersiltigsandigerkiesauffullung",
                    "rect": [
                        232.78799438476562,
                        130.18496704101562,
                        525.6640014648438,
                        153.54295349121094
                    ],
                    "lines": [
                        {
                            "text": "grauer, siltig-sandiger Kies (Auffullung)",
                            "rect": [
                                232.78799438476562,
                                130.18496704101562,
                                525.6640014648438,
                                153.54295349121094
                            ],
                            "page": 1
                        }
                    ],
                    "page": 1
                },
                "depth_interval": {
                    "start": null,
                    "end": {
                        "value": 0.4,
                        "rect": [
                            125.25399780273438,
                            140.2349853515625,
                            146.10398864746094,
                            160.84498596191406
                        ],
                        "page": 1
                    }
                }
            },
...
            {
                "material_description": {
                    "text": "dunkelgrauermagererztsandigerlehm",
                    "rect": [
                        246.42300415039062,
                        708.9439697265625,
                        438.72283935546875,
                        722.6839599609375
                    ],
                    "lines": [
                        {
                            "text": "dunkelgrauer, magerer, z.T. sandiger Lehm",
                            "rect": [
                                246.42300415039062,
                                708.9439697265625,
                                438.72283935546875,
                                722.6839599609375
                            ],
                            "page": 1
                        }
                    ],
                    "page": 1
                },
                "depth_interval": {
                    "start": {
                        "value": 9.9,
                        "rect": [
                            163.56900024414062,
                            684.2750244140625,
                            177.468994140625,
                            698.0150146484375
                        ],
                        "page": 1
                    },
                    "end": {
                        "value": 12.0,
                        "rect": [
                            160.0050048828125,
                            745.1210327148438,
                            177.5189971923828,
                            757.4869995117188
                        ],
                        "page": 1
                    }
                }
            }
        ],
        "depths_materials_column_pairs": [
            {
                "depth_column": {
                    "rect": [
                        160.0050048828125,
                        410.78302001953125,
                        178.76199340820312,
                        757.4869995117188
                    ],
                    "entries": [
                        {
                            "value": 0.2,
                            "rect": [
                                163.0070037841797,
                                410.78302001953125,
                                176.90699768066406,
                                424.52301025390625
                            ],
                            "page": 1
                        },
                        {
                            "value": 2.9,
                            "rect": [
                                162.36900329589844,
                                485.98602294921875,
                                176.2689971923828,
                                499.72601318359375
                            ],
                            "page": 1
                        },
                        {
                            "value": 3.1,
                            "rect": [
                                163.23500061035156,
                                495.28399658203125,
                                177.13499450683594,
                                509.02398681640625
                            ],
                            "page": 1
                        },
                        {
                            "value": 3.7,
                            "rect": [
                                163.31199645996094,
                                509.58197021484375,
                                177.2119903564453,
                                523.3219604492188
                            ],
                            "page": 1
                        },
                        {
                            "value": 4.0,
                            "rect": [
                                163.27000427246094,
                                517.2839965820312,
                                177.1699981689453,
                                531.0239868164062
                            ],
                            "page": 1
                        },
                        {
                            "value": 4.3,
                            "rect": [
                                163.1320037841797,
                                526.6389770507812,
                                177.03199768066406,
                                540.3789672851562
                            ],
                            "page": 1
                        },
                        {
                            "value": 5.1,
                            "rect": [
                                163.4720001220703,
                                548.4299926757812,
                                178.76199340820312,
                                563.5440063476562
                            ],
                            "page": 1
                        },
                        {
                            "value": 5.7,
                            "rect": [
                                163.5570068359375,
                                567.0780029296875,
                                177.45700073242188,
                                580.8179931640625
                            ],
                            "page": 1
                        },
                        {
                            "value": 7.0,
                            "rect": [
                                163.42300415039062,
                                602.1010131835938,
                                177.322998046875,
                                615.8410034179688
                            ],
                            "page": 1
                        },
                        {
                            "value": 7.7,
                            "rect": [
                                163.625,
                                622.0120239257812,
                                177.52499389648438,
                                635.7520141601562
                            ],
                            "page": 1
                        },
                        {
                            "value": 8.6,
                            "rect": [
                                163.5500030517578,
                                647.9769897460938,
                                177.4499969482422,
                                661.7169799804688
                            ],
                            "page": 1
                        },
                        {
                            "value": 9.9,
                            "rect": [
                                163.56900024414062,
                                684.2750244140625,
                                177.468994140625,
                                698.0150146484375
                            ],
                            "page": 1
                        },
                        {
                            "value": 12.0,
                            "rect": [
                                160.0050048828125,
                                745.1210327148438,
                                177.5189971923828,
                                757.4869995117188
                            ],
                            "page": 1
                        }
                    ]
                },
                "material_description_rect": [
                    244.97000122070312,
                    414.0959777832031,
                    594.2720947265625,
                    722.6839599609375
                ]
            }
        ],
        "page_height": [
            1192.0
        ],
        "page_width": [
            842.0
        ]
    }
}
```

Behind the scenes, in the code, the page-based logic was removed, and a File-based logic was introduced. 

Additional features: 
- Added the `launch.json` file that makes it possible to run the borehole extraction in debug mode. 
- Added docstrings to some public functions

### CODE REVIEW

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
I pushed two commits:
- one fixing 2 TODOs in predictions.py (many thanks for highlighting those!) 
- one addressing the last point in the JIRA ticket:
  - _While we are at it, can we remove the postprocessing (lowercase, removing whitespace) from the "text" field of the material description (e.g. "text": "kiesmitvielsandwenigsiltundtonbraungrau")? It would make more sense to me if this postprocessing is only done for the purposes of evaluation / comparing with ground truth, but the predictions.json file should actually contain the original text from the document without postprocessing (as is already the case for the individual lines)._

Remaining points from the ticket:
- _Importing the extracted data in Label Studio still works_
  - Did you check this @dcleres ?
- _README documentation is updated_
  - The README currently contains a JSON example with the old structure

I'm not a big fan of the fact that we had to add `page_number=1` in so many unit tests that are not really concerned with page number logic at all. However, I don't really see an immediate solution to this. Unless you see an easy way to avoid this, I'm happy to accept it like this.

The other change requests that I've added are all pretty minor, I think.

File: src/stratigraphy/util/coordinate_extraction.py [hunk 1]
```diff
@@ -312,8 +312,9 @@ def extract_coordinates(self) -> Coordinate | None:
         Returns:
             Coordinate | None: the extracted coordinates (if any)
         """
-        for page in self.doc:
-            lines = extract_text_lines(page)
+        for page_idx, page in enumerate(self.doc):
+            page_number = page_idx + 1
+            lines = extract_text_lines(page, page_number)
             page_number = page.number + 1  # page.number is 0-based
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/util/coordinate_extraction.py:318]: Good catch. thank you. I do not know how I missed that.

File: src/stratigraphy/main.py [hunk 2]
```diff
@@ -208,11 +214,18 @@ def start_pipeline(
                                 )
                                 mlflow.log_image(img, f"pages/{filename}_page_{page.number + 1}_lines.png")
 
+                    predictions[filename]["layers"] = layer_predictions_list
+                    predictions[filename]["depths_materials_column_pairs"] = depths_materials_column_pairs_list
+                    predictions[filename]["page_height"] = page_heights_list
+                    predictions[filename]["page_width"] = page_widths_list
+
+                    assert len(page_heights_list) == len(page_widths_list) == doc.page_count, "Page count mismatch."
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/main.py:222]: I am fine with that.

File: src/stratigraphy/util/find_depth_columns.py [hunk 3]
```diff
@@ -32,10 +33,10 @@ def depth_column_entries(all_words: list[TextWord], include_splits: bool) -> lis
             match = regex.match(input_string)
             if match:
                 value = value_as_float(match.group(1))
-                entries.append(DepthColumnEntry(word.rect, value))
+                entries.append(DepthColumnEntry(word.rect, value, page_number))
             elif include_splits:
                 # support for e.g. "1.10-1.60m" extracted as a single word
-                layer_depth_column_entry = extract_layer_depth_interval(input_string, word.rect)
+                layer_depth_column_entry = extract_layer_depth_interval(input_string, word.rect, page_number)
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/util/find_depth_columns.py:39]: Thank you I missed the fact that the word knows its page.

File: src/stratigraphy/util/predictions.py [hunk 4]
```diff
@@ -36,31 +36,25 @@ class LayerPrediction:
...
+    def __init__(
+        self,
+        layers: list[LayerPrediction],
+        file_name: str,
+        language: str,
+        metadata: BoreholeMetaData = None,
+        depths_materials_columns_pairs: list[dict] = None,
+        page_sizes: list[tuple[int, int]] = None,
+    ):
+        self.layers: list[LayerPrediction] = sorted(layers, key=lambda layer: layer.material_description.rect.y0)
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/util/predictions.py:51]: I think during development, it helped me to better find the text on the page and make a few visual tests. However, you are correct that this does not impact the performance of the extraction. So, I will get back to the original version. Thank you for highlighting this.

**Comment by dcleres** [comment]
I found during the review that the file: `src/scripts/label_studio_annotation_to_ground_truth.py` also needs to be adapted to match the new format. I will start working on this.

**Comment by dcleres** [comment]
Thank you for your comment regarding the PR. I resolved all the issues you highlighted which are relevant to this part of the code. The changes required to the label studio repository will be merged into the respective repository. 

> I'm not a big fan of the fact that we had to add page_number=1 in so many unit tests that are not really concerned with page number logic at all. However, I don't see an immediate solution to this. Unless you see an easy way to avoid this, I'm happy to accept it as is.

Regarding this comment, I fully agree with you, but I also have not found a fundamentally better way to solve this issue. I edited the tests and added a constant called `PAGE_NUMBER` that mitigates this issue to some extent. 

While doing this, I also realized that we are not doing any tests on files that have more than one page. Should a test be added for this as well?

**Comment by stijnvermeeren-swisstopo** [comment]
> While doing this, I also realized that we are not performing any tests on files with more than one page. @stijnvermeeren-swisstopo, should a test be added for this as well?

That would certainly be useful as well, but it's not a top priority. If you want to work on that, I would certainly do it in a separate PR, so that we can close this one already now.

**Review by stijnvermeeren-swisstopo — APPROVED** [review]
I pushed a minor simplification of the page dimensions code (https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/70/commits/4404eff5b61c6827d054bc000e406aedb3f5ab32).

Everything looks good to me now.

**Pull request merged**